### PR TITLE
yl2x is a DigitalMars-specific builtin

### PR DESCRIPTION
**Pull request by ibuclaw**
Fix linker errors building with gdc, much more to come... :-)

### CODE REVIEW

**Comment by WalterBright** [comment]
The y2lx instruction is x86 specific, not Digital Mars specific.

**Comment by ibuclaw** [comment]
Is ``INLINE_YL2X`` set in the DMD compiler?  That would be one compromise if it did.

**Comment by WalterBright** [comment]
> Is  INLINE_YL2X  set in the DMD compiler?

No. Just provide a yl2x function in the GDC port.

**Comment by ibuclaw** [comment]
> No. Just provide a yl2x function in the GDC port.

I'd rather not.  There is no way to map it to any architecture (even x86) on my side.

**Comment by WalterBright** [comment]
Why wouldn't inline assembler work?

**Comment by ibuclaw** [comment]
It goes against convention - and will be a red flag for any reviewers in upstream GCC.

It's difficult enough fixing the front-end so it doesn't care about what target it is generating code for, let alone what *host* we will be running on.

**Comment by yebblies** [comment]
This change means that ddmd built with gdc won't be able to evaluate the yl2x intrinsics at compile time, while dmd built with gcc can?  Does gcc/gdc really not provide this builtin in some way?

**Comment by yebblies** [comment]
Sorry, hit wrong button.

**Comment by ibuclaw** [comment]
> Sorry, hit wrong button.

You're lucky I'll be absent next week, or you will have something to be sorry about... :-)

> This change means that ddmd built with gdc won't be able to evaluate the yl2x intrinsics at compile time, while dmd built with gcc can? Does gcc/gdc really not provide this builtin in some way?

There's log2l, but it will produce wildly different results from ``fy2lx`` - that is to say, dmd's fy2lx will give incorrectly rounded results vs. log2l.

**Comment by yebblies** [comment]
> You're lucky I'll be absent next week, or you will have something to be sorry about... :-)

You're not coming this year?

> There's log2l, but it will produce wildly different results from fy2lx - that is to say, dmd's fy2lx will give incorrectly rounded results vs. log2l.

Hmm ok.  Do we have a ctfeable implementation in core.math?  If so, can we just completely remove ctfe builtin support?  It seems like much too obscure a thing to have compiler code dedicated to it.

I don't want a compiler built by dmd and a compiler built by gdc to have different capabilities, I'm sure you can agree.

**Comment by ibuclaw** [comment]
>  Do we have a ctfeable implementation in core.math?

No.

> It seems like much too obscure a thing to have compiler code dedicated to it

This is what I said in the original PR for CTFE-yl2x!  You still merged it in despite my warnings.

> I don't want a compiler built by dmd and a compiler built by gdc to have different capabilities

On that note, ``std.math.tan`` is a problematic function too that DMD treats like a built-in.

**Comment by yebblies** [comment]
>> Do we have a ctfeable implementation in core.math?

> No.

Oh yeah, it doesn't need to be ctfeable.  Just present.  Does GDC druntime not provide this either?

> This is what I said in the original PR for CTFE-yl2x! You still merged it in despite my warnings.

I didn't merge it.

> On that note, `std.math.tan` is a problematic function too that DMD treats like a built-in.

What's so problematic about `tan`?

But yeah, we should probably avoid adding ctfe intrinsics that depend on the host compiler/runtime.

**Comment by ibuclaw** [comment]
> What's so problematic about tan?

It's a weird half/half.  DMD recognises it as a CTFE-able builtin, but runtime it uses inline assembler.

For simplicity on my part, I only implement body-less functions in std.math/core.math as intrinsics.

**Comment by ibuclaw** [comment]
And... I found a case where the implied use of yl2x broke gdc, along with every non-x86 target.  https://github.com/D-Programming-Language/phobos/pull/3309

*Sigh*  -  We should put a stop to this by removing uses of yl2x from std.math.  They might have been useful in the beginning.  But not anymore.

**Comment by yebblies** [comment]
Or by implementing a portable version somewhere.

**Comment by ibuclaw** [comment]
> Or by implementing a portable version somewhere.

It's caled ``log2``, I did this 2 years ago.

https://github.com/D-Programming-Language/phobos/commit/bf2b2a1858000dc88f0f3940faf0556feb4a90df

File: src/root/port.d [hunk 1]
```diff
@@ -149,12 +157,14 @@ extern (C++) struct Port
 
     static void yl2x_impl(real* x, real* y, real* res)
     {
-        *res = yl2x(*x, *y);
+        version(DigitalMars)
+            *res = yl2x(*x, *y);
```
Comment (MartinNowak) [inline src/root/port.d]: else assert(0);

**Pull request merged**